File: src/canvas/membraneStretch.ts
// ────────────────────────────────────────────────────────────────────────────
// MEMB-4 — Le mode étiré, à l'échelle d'un board.
//
// PROBLÈME. `stretchPlan` (dans `membraneSpace`) sait déjà, pour UNE membrane,
// jusqu'où elle a le droit de pousser et qui l'en empêche. Il lui manquait un
// appelant, et cet appelant doit trancher deux questions que la géométrie seule
// ne pose pas :
//
//   ① QUI compte comme obstacle ?
//   ② DANS QUEL ORDRE traiter des membranes imbriquées ?
//
// ① LES OBSTACLES SONT LES FRÈRES, ET RIEN D'AUTRE. `stretchPlan` compare des
// boîtes NATURELLES. Or deux éléments n'ont de coordonnées comparables que
// s'ils vivent dans le même repère — c'est-à-dire s'ils ont la même membrane
// parente. Un élément posé au fond d'une AUTRE membrane a des coordonnées
// naturelles qui ne disent rien de sa place à l'écran (sa membrane peut le
// réduire de moitié) : le confronter à notre boîte comparerait deux repères
// différents. Ce n'est pas une perte : cette autre membrane est elle-même un
// frère, et c'est SA boîte qui fait obstacle — au bon niveau, dans le bon
// repère. Cette règle exclut gratuitement les trois cas qu'il aurait fallu
// écarter à la main : la membrane elle-même, sa descendance (qui grandit avec
// elle) et ses ancêtres (qui la contiennent, donc la chevauchent toujours).
//
// ② ON TRAITE LES PLUS PROFONDES D'ABORD. Une membrane étirée à l'intérieur
// d'une autre grandit, et sa nouvelle boîte fait partie du contenu que la
// parente doit contenir. Mesurer la parente avant que l'enfant ait poussé la
// ferait grandir d'un cran de retard — visible à l'usage comme un « il faut
// bouger deux fois pour que ça s'ajuste ». On descend donc par profondeur
// décroissante, en reportant chaque croissance dans la copie de travail.
//
// CE QUI RESTE VRAI D'AILLEURS. On applique `allowed`, JAMAIS `desired` : la
// membrane bute sur l'obstacle, elle ne le recouvre pas et ne le capture pas
// (décision produit). Et une membrane ne RÉTRÉCIT jamais toute seule — un
// utilisateur qui l'a agrandie à la main a exprimé une intention que retirer du
// contenu ne doit pas effacer.
//
// QUAND C'EST APPELÉ. À des instants DISCRETS — fin d'un glisser, conversion de
// mode — jamais en continu. Même raison que pour l'appartenance (cf. l'en-tête
// de `membraneSpace`) : une géométrie qui se réécrit à chaque frame de drag
// noierait l'annulation et ferait fuir la cible sous le curseur.
// ────────────────────────────────────────────────────────────────────────────

import { MEMBRANE_SPACE, parentMap, stretchPlan, type SpaceItem } from "./membraneSpace";

/** Ce qu'une membrane étirée doit devenir, et ce qui l'a empêchée d'aller plus loin. */
export interface StretchOutcome {
  membraneId: string;
  /** Taille à écrire — c'est `allowed`, jamais `desired`. */
  width: number;
  height: number;
  /** La boîte a-t-elle effectivement changé ? */
  grew: boolean;
  /** La croissance a-t-elle été rabotée par un obstacle ? */
  blocked: boolean;
  /** Ids des éléments à entourer, dans l'ordre de rencontre. */
  blockerIds: string[];
}

/** Profondeur d'imbrication d'un élément (0 = à la racine du board). */
function depthOf(id: string, parents: Map<string, string>): number {
  const seen = new Set<string>([id]);
  let d = 0;
  let cur = parents.get(id);
  while (cur && !seen.has(cur)) {
    seen.add(cur);
    d += 1;
    cur = parents.get(cur);
  }
  return d;
}

/**
 * Croissance de TOUTES les membranes étirées d'un board.
 *
 * Fonction pure : `items` n'est pas modifié (on travaille sur une copie), et
 * seules les membranes qui bougent ou qui butent sont rendues — une membrane
 * étirée déjà à la bonne taille et sans obstacle ne produit rien, donc
 * n'entraîne aucune écriture dans le document.
 */
export function planBoardStretch(items: SpaceItem[]): StretchOutcome[] {
  const parents = parentMap(items);
  const work: SpaceItem[] = items.map((i) => ({ ...i }));
  const byId = new Map(work.map((i) => [i.id, i]));

  const targets = work
    .filter((i) => i.kind === "membrane" && i.mode === "stretched")
    .sort((a, b) => depthOf(b.id, parents) - depthOf(a.id, parents));

  const out: StretchOutcome[] = [];
  for (const m of targets) {
    const frame = parents.get(m.id) ?? null;
    const children: SpaceItem[] = [];
    const foreigners: SpaceItem[] = [];
    for (const it of work) {
      if (it.id === m.id) continue;
      const p = parents.get(it.id) ?? null;
      if (p === m.id) children.push(it);
      else if (p === frame) foreigners.push(it);
    }

    const plan = stretchPlan(m, children, foreigners);
    const width = plan.allowed.width;
    const height = plan.allowed.height;
    const grew = width !== m.width || height !== m.height;
    if (!grew && !plan.blocked) continue;

    // Reporté dans la copie de travail : la membrane parente, traitée juste
    // après, doit mesurer la boîte que celle-ci vient de prendre.
    const live = byId.get(m.id);
    if (live) { live.width = width; live.height = height; }

    out.push({
      membraneId: m.id,
      width, height, grew,
      blocked: plan.blocked,
      blockerIds: plan.blockers.map((b) => b.id),
    });
  }
  return out;
}

/** Marge laissée entre le contenu et le bord — réexportée pour l'interface. */
export const STRETCH_PADDING = MEMBRANE_SPACE.STRETCH_PADDING;

File: crates/glucose-core/src/membrane_stretch.rs
//! MEMB-4 — Le mode étiré, à l'échelle d'un board (géométrie PURE, 0 dépendance).

use crate::geometry::Rect;
use crate::membrane_space::{content_extent, parent_map, SpaceItem, STRETCH_PADDING};
use crate::types::MembraneMode;
use std::collections::{HashMap, HashSet};

#[derive(Debug, Clone, PartialEq)]
pub struct StretchPlan {
    pub desired: Rect,
    pub allowed: Rect,
    pub blockers: Vec<SpaceItem>,
    pub blocked: bool,
}

pub fn stretch_plan(
    membrane: &SpaceItem,
    children: &[&SpaceItem],
    foreigners: &[&SpaceItem],
) -> StretchPlan {
    let m_rect = membrane.rect();
    let (ext_w, ext_h) = content_extent(m_rect, children);
    let pad = STRETCH_PADDING;
    let desired = Rect::new(
        membrane.x,
        membrane.y,
        membrane.width.max(ext_w + pad),
        membrane.height.max(ext_h + pad),
    );

    let mut max_w = desired.width;
    let mut max_h = desired.height;
    let mut blockers = Vec::new();

    for &f in foreigners {
        if f.id == membrane.id {
            continue;
        }
        let f_rect = f.rect();
        if !desired.overlaps(f_rect) {
            continue;
        }
        if m_rect.overlaps(f_rect) {
            continue;
        }

        blockers.push(f.clone());
        let cut_w = f.x - membrane.x;
        let cut_h = f.y - membrane.y;
        let lose_w = desired.width - cut_w;
        let lose_h = desired.height - cut_h;
        if lose_w <= lose_h {
            max_w = max_w.min(membrane.width.max(cut_w));
        } else {
            max_h = max_h.min(membrane.height.max(cut_h));
        }
    }

    let allowed = Rect::new(membrane.x, membrane.y, max_w, max_h);
    let blocked = allowed.width < desired.width || allowed.height < desired.height;

    StretchPlan {
        desired,
        allowed,
        blockers,
        blocked,
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct StretchOutcome {
    pub membrane_id: String,
    pub width: f64,
    pub height: f64,
    pub grew: bool,
    pub blocked: bool,
    pub blocker_ids: Vec<String>,
}

fn depth_of(id: &str, parents: &HashMap<String, String>) -> usize {
    let mut seen = HashSet::new();
    seen.insert(id);
    let mut d = 0;
    let mut cur = parents.get(id);
    while let Some(c) = cur {
        if seen.contains(c.as_str()) {
            break;
        }
        seen.insert(c.as_str());
        d += 1;
        cur = parents.get(c);
    }
    d
}

/// Croissance de TOUTES les membranes étirées d'un board (par profondeur décroissante).
pub fn plan_board_stretch(items: &[SpaceItem]) -> Vec<StretchOutcome> {
    let parents = parent_map(items);
    let mut work: Vec<SpaceItem> = items.to_vec();

    let mut target_indices: Vec<usize> = work
        .iter()
        .enumerate()
        .filter(|(_, i)| {
            i.kind == crate::membrane_space::SpaceItemKind::Membrane
                && i.mode == Some(MembraneMode::Stretched)
        })
        .map(|(idx, _)| idx)
        .collect();

    target_indices.sort_by(|&a, &b| {
        let da = depth_of(&work[a].id, &parents);
        let db = depth_of(&work[b].id, &parents);
        db.cmp(&da)
    });

    let mut out = Vec::new();
    for idx in target_indices {
        let m = work[idx].clone();
        let frame = parents.get(&m.id).cloned();

        let mut children = Vec::new();
        let mut foreigners = Vec::new();

        for it in &work {
            if it.id == m.id {
                continue;
            }
            let p = parents.get(&it.id).cloned();
            if p.as_deref() == Some(&m.id) {
                children.push(it);
            } else if p == frame {
                foreigners.push(it);
            }
        }

        let plan = stretch_plan(&m, &children, &foreigners);
        let width = plan.allowed.width;
        let height = plan.allowed.height;
        let grew = width != m.width || height != m.height;
        if !grew && !plan.blocked {
            continue;
        }

        work[idx].width = width;
        work[idx].height = height;

        out.push(StretchOutcome {
            membrane_id: m.id,
            width,
            height,
            grew,
            blocked: plan.blocked,
            blocker_ids: plan.blockers.into_iter().map(|b| b.id).collect(),
        });
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::membrane_space::SpaceItemKind;

    #[test]
    fn test_stretch_plan_unblocked() {
        let m = SpaceItem {
            id: "M1".into(),
            kind: SpaceItemKind::Membrane,
            x: 0.0,
            y: 0.0,
            width: 100.0,
            height: 100.0,
            mode: Some(MembraneMode::Stretched),
            membrane_id: None,
        };
        let kid = SpaceItem {
            id: "K1".into(),
            kind: SpaceItemKind::Image,
            x: 20.0,
            y: 20.0,
            width: 150.0,
            height: 150.0,
            mode: None,
            membrane_id: Some("M1".into()),
        };

        let plan = stretch_plan(&m, &[&kid], &[]);
        assert!(!plan.blocked);
        assert!(plan.allowed.width >= 170.0 + STRETCH_PADDING);
        assert!(plan.allowed.height >= 170.0 + STRETCH_PADDING);
    }

    #[test]
    fn test_stretch_plan_blocked_by_foreigner() {
        let m = SpaceItem {
            id: "M".into(),
            kind: SpaceItemKind::Membrane,
            x: 0.0,
            y: 0.0,
            width: 200.0,
            height: 200.0,
            mode: Some(MembraneMode::Stretched),
            membrane_id: None,
        };
        let kid = SpaceItem {
            id: "I".into(),
            kind: SpaceItemKind::Image,
            x: 0.0,
            y: 0.0,
            width: 300.0,
            height: 100.0,
            mode: None,
            membrane_id: Some("M".into()),
        };
        let obstacle = SpaceItem {
            id: "ETR".into(),
            kind: SpaceItemKind::Image,
            x: 250.0,
            y: 0.0,
            width: 50.0,
            height: 50.0,
            mode: None,
            membrane_id: None,
        };

        let plan = stretch_plan(&m, &[&kid], &[&obstacle]);
        assert!(plan.blocked);
        assert_eq!(plan.blockers.len(), 1);
        assert_eq!(plan.blockers[0].id, "ETR");
        assert_eq!(plan.allowed.width, 250.0);
    }
}
